### PR TITLE
Added get_proj_plot() method in phonon.plotter.PhononBSPlotter class

### PR DESCRIPTION
**Pull request by cnncnnzh**
## Summary

A method that allows users to plot the atom-resolved phonon band structures. The contribution from each type (or group) of atoms is calculated based on the magnitude of the eigendisplacements.  Support 2 (rb), 3 (rgb) and 4 (cmyk) unique groups of atoms.

So far I just used the existing NaCl example for testing, let me know if more cases are needed!

## Additional dependencies introduced (if any)

matplotlib.pyplot.plt —> essential for the colored plot
matplotlib.collections.LineCollection —> essential for the colored plot
pymatgen.electronic_structure.plotter.BSDOSPlotter —> For creating the color line and color triangle

## TODO (if any)

1. Make a color legend for the 4-color case.
2. Allow projecting band based on eigenvectors instead of eigendisplacement?

I would appreciate any suggestions from you! Thanks.

### CODE REVIEW

File: pymatgen/phonon/plotter.py [hunk 1]
```diff
@@ -739,6 +867,7 @@ def get_plot(self, marker="o", markersize=None, units="thz"):
             units: unit for the plots, accepted units: thz, ev, mev, ha, cm-1, cm^-1
 
         Returns: plot
+
```
Comment (janosh) [inline pymatgen/phonon/plotter.py:870]: Could you undo the white space changes?

File: pymatgen/phonon/plotter.py [hunk 2]
```diff
@@ -381,6 +381,134 @@ def get_plot(self, ylim=None, units="thz"):
...
+        elif len(colors) == 3:
+            return colors
+        # if there are four groups, use cyan, magenta, yellow and black
+        elif len(colors) == 4:
+            r = (1 - colors[0]) * (1 - colors[3])
+            g = (1 - colors[1]) * (1 - colors[3])
+            b = (1 - colors[2]) * (1 - colors[3])
+            return [r, g, b]
+
+    def get_proj_plot(self, site_comb="element", ylim=None, units="thz"):
```
Comment (janosh) [inline pymatgen/phonon/plotter.py:425]: Would be great if new code had type annotations. Something like:

```py
    def get_proj_plot(self, site_comb: str = "element", ylim: tuple[float | None, float | None] = None, units: str ="thz") -> plt.Axes:
```

**Review by janosh — CHANGES_REQUESTED** [review]
Looks good! 👍 Just a few nitpicks.

File: pymatgen/phonon/tests/test_plotter.py [hunk 3]
```diff
@@ -77,6 +77,13 @@ def test_plot(self):
         rc("text", usetex=False)
         self.plotter.get_plot(units="mev")
 
+    def test_proj_plot(self):
+        # Disabling latex for testing.
+        from matplotlib import rc
+
+        rc("text", usetex=False)
+        self.plotter.get_proj_plot(units="mev")
```
Comment (janosh) [inline pymatgen/phonon/tests/test_plotter.py:85]: Try to test all (or most) code paths, i.e. pass different values for all arguments (incl. `site_comb` and `ylim`) to test edge cases.

Comment (cnncnnzh) [reply pymatgen/phonon/tests/test_plotter.py:89]: It's my first-time contributing, so thank you for all the suggestions. Here are some changes I made:

- delete the extra space.
- added type hints.
- added more assertions to catch exceptions (e.g., specify more atoms than what the structure has, not list all atoms, etc.).
- added a new test file SrTiO3_phonon_bandstructure.json to test len(site_comb) = 3 or 4 cases.
- pass different values to get_proj_plot().

Let me know if there is anything else. Thanks!

**Comment by janosh** [comment]
Here's the plot output from `test_proj_plot()` for later reference/discussion.

Is the text extending beyond the legend boundary easy to fix?

test_proj_plot-1
test_proj_plot-2
test_proj_plot-3
test_proj_plot-4
test_proj_plot-5
test_proj_plot-6
test_proj_plot-7

**Comment by cnncnnzh** [comment]
I changed the default legends to "0", "1" and "2" so that they fit in the box. But now the rgb_label variable can be passed to the method to change the legend. How does that sound?

Figure 2022-10-12 222115

**Comment by janosh** [comment]
I think you force-pushed over the commit 709cda1 I added after 702b5bd to fix `mypy` errors. Would be better if you could `git pull` remote changes to your branch first before you force push in the future.

Same thing with the commits pushed by the `pre-commit.ci` bot. Simply pull them in with `git pull`.

> I changed the default legends to "0", "1" and "2" so that they fit in the box. But now the rgb_label variable can be passed to the method to change the legend. How does that sound?

Sounds good. 👍

File: pymatgen/phonon/plotter.py [hunk 4]
```diff
@@ -407,6 +548,46 @@ def save_plot(self, filename, img_format="eps", ylim=None, units="thz"):
...
+        self,
+        filename,
+        img_format="eps",
+        site_comb: str | list[list[int]] = "element",
+        ylim: tuple[None | float, None | float] = None,
+        units: str = "thz",
+        rgb_labels: tuple[str] = None,
+    ):
+        """
+        Save matplotlib projected plot to a file.
```
Comment (janosh) [inline pymatgen/phonon/plotter.py:579]: One more question: do we really need `save_proj_plot()`? Seems like it's a wrapper around calling `fig.savefig()` which can just be left to the user.

Comment (cnncnnzh) [reply pymatgen/phonon/plotter.py:579]: > I think you force-pushed over the commit [709cda1](https://github.com/materialsproject/pymatgen/commit/709cda100218e47dba247ba90aadf649e1d7912b) I added after [702b5bd](https://github.com/materialsproject/pymatgen/commit/702b5bdc8e054b0f269dffadc44440f3098bf16b) to fix `mypy` errors. Would be better if you could `git pull` remote changes to your branch first before you force push in the future.
> 
> Same thing with the commits pushed by the `pre-commit.ci` bot. Simply pull them in with `git pull`.
> 
> > I changed the default legends to "0", "1" and "2" so that they fit in the box. But now the rgb_label variable can be passed to the method to change the legend. How does that sound?
> 
> Sounds good. 👍

Oh sure. I'll do that in the future 👍.

Comment (cnncnnzh) [reply pymatgen/phonon/plotter.py:579]: > One more question: do we really need `save_proj_plot()`? Seems like it's a wrapper around calling `fig.savefig()` which can just be left to the user.

I intended to write a counterpart for 'save_plot()'. But I think you are right that it's not necessary.

**Review by janosh — APPROVED** [review]
Thanks for this contribution! 👍
